doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,456 @@
+transaction happen at the same timestamp. All writes that happen after the first read of a read-only transaction will
+happen at a later timestamp, and therefore will not affect any reads of the read-only transaction.
+
+Informally we can use the transaction timestamps to provide a total ordering over all transactions.
+
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 < the timestamp of T2.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is write-only, and
+  T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  read-only, and T2 is read-only.
+
+#### Strict Serializable
+
+See [https://jepsen.io/consistency/models/strict-serializable](https://jepsen.io/consistency/models/strict-serializable)
+for an in depth formal discussion of Strict Serializable. Strict Serializable is a combination of Serializable and
+Linearizable, which means that the total order formed by all transactions must be constrained by their real time
+occurrences. So if transaction t1 happened in real time before transaction t2, then t1 must be ordered before t2 in the
+total ordering.
+
+Operations that only include user tables, and include at least one user table, are currently Strict Serializable. This
+is accomplished by using a single global timestamp and a single thread in the coordinator to order operations. Writes
+happen at a timestamp larger than all previous operations. Reads happen at a timestamp equal to or larger than all
+previous operations.
+
+Operations that involve reading from views over upstream data are not currently Strict Serializable. It is possible to
+read from a view at one timestamp and then later read from the same view at an earlier timestamp. Consider the following
+scenario:
+
+1. Client reads from view v1 that has an upper of `ts + 1` and a since of `ts - 10`. The read will happen at `ts`.
+2. Client reads from a join over v1 and view v2. v2 has an upper of `ts - 5` and some since less or equal to `ts - 6`.
+   The read will happen at `ts - 6`.
+
+Reference [Timestamp Selection](#Timestamp Selection) to see why these timestamps are selected. This is an obvious
+violation of strict serializability because the second read happens later than the first read in real time, but it is
+assigned an earlier timestamp.
+
+In order to achieve Strict Serializability in the general case, Materialize must always select a timestamp greater than
+or equal to all previously selected timestamps. This will ensure that if an operation happens later in real time, then
+it will be assigned an equal or later timestamp. Consecutive reads can (but don't have to) all be assigned an equal
+timestamp as long as there’s no writes in between them. Consecutive writes can (but don't have to) all be assigned an
```

Comment:
Is this still true? Writes only get timestamped at commit, and each separate commit must have a distinct timestamp. Or maybe I'm being too pedantic about what "consecutive" means here; I read that as "not in the same transaction" (those in the same transaction would be "concurrent")

Reply:
I was referring to consecutive write transactions, so you read it correctly. I was trying to differentiate between "can" and "must" here. In our current implementation consecutive writes do get increasing timestamps, but that's not a requirement for  consistency it's just an artifact of our current implementation. There are potential implementations we could use that assign many write transactions the same timestamp.